### ISSUE TITLE
ray.exceptions:RpcError & KVStoreError

### ISSUE THREAD

**Issue by rsong0606** [ray-project/ray#45580]
### What happened + What you expected to happen

Hey Team,
Good work! I got the same issue as this ticket 
https://github.com/ray-project/ray/issues/41657
This ticket has been closed, but no clear suggestions gave. Can someone take a look and share how we can solve it?

For my case, I cascaded or combined all models in one application call
```
my_fastapi_deployment = MyFastAPIDeployment.bind(
    dictionary_lookup, image_handler, chatbot,
    audio_evaluator, dictionary_viewer, quiz_handler, audio_generator
)
```
In above call, I have a gptq llama2 chatbot model, a small TTS model, one image to text model roughly taking 17gb memory in total.

```
import os
import sys
from pathlib import Path

# Add the project directory to the Python path
PROJECT_DIR = Path(__file__).resolve().parent
sys.path.append(str(PROJECT_DIR))

import ray
from ray import serve
from ray_deploy import dictionary_lookup, image_handler, chatbot, audio_evaluator, dictionary_viewer, quiz_handler, audio_generator
from fast_api import app, MyFastAPIDeployment  # Correctly import the FastAPI app and deployment class
# Initialize Ray and Ray Serve
ray.init(address="auto")
serve.start()




my_fastapi_deployment = MyFastAPIDeployment.bind(
    dictionary_lookup, image_handler, chatbot,
    audio_evaluator, dictionary_viewer, quiz_handler, audio_generator
)
serve.run(my_fastapi_deployment)

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)


```
and I runs into the error message exactly as the same issue in another ticket provided above
```
       serve.run(my_fastapi_deployment)
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/api.py", line 578, in run
    handle = _run(
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/api.py", line 529, in _run
    client.deploy_application(
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/_private/client.py", line 45, in check
    return f(self, *args, **kwargs)
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/_private/client.py", line 278, in deploy_application
    ray.get(self._controller.deploy_application.remote(name, deployment_args_list))
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/_private/auto_init_hook.py", line 21, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/_private/worker.py", line 2623, in get
    values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/_private/worker.py", line 861, in get_objects
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(KVStoreError): ray::ServeController.deploy_application() (pid=4655, ip=10.128.0.34, actor_id=b9c62c93a5c11aeffec6711a01000000, repr=<ray.serve._private.controller.ServeController object at 0x7f21ab9878e0>)
ray.exceptions.RpcError: Sent message larger than max (627350011 vs. 536870912)

During handling of the above exception, another exception occurred:

ray::ServeController.deploy_application() (pid=4655, ip=10.128.0.34, actor_id=b9c62c93a5c11aeffec6711a01000000, repr=<ray.serve._private.controller.ServeController object at 0x7f21ab9878e0>)
  File "/opt/conda/envs/myenv/lib/python3.9/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/opt/conda/envs/myenv/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/_private/controller.py", line 749, in deploy_application
    self.application_state_manager.deploy_app(name, deployment_args_deserialized)
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/_private/application_state.py", line 858, in deploy_app
    self._application_states[name].deploy_app(deployment_infos)
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/_private/application_state.py", line 376, in deploy_app
    self._set_target_state(
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/_private/application_state.py", line 272, in _set_target_state
    self._save_checkpoint_func(writeahead_checkpoints={self._name: target_state})
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/_private/application_state.py", line 1000, in _save_checkpoint_func
    self._kv_store.put(
  File "/opt/conda/envs/myenv/lib/python3.9/site-packages/ray/serve/_private/storage/kv_store.py", line 67, in put
    raise KVStoreError(e.rpc_code)
ray.serve._private.storage.kv_store.KVStoreError: 8

```

### Versions / Dependencies

ray version 2.23.0
pytorch
vllm
transformers
accelerate
fastapi

### Reproduction script

Unfortunately, there's lots of dependency to make it reproducible. Any comments will be greatly appreciated!

### Issue Severity

High: It blocks me from completing my task.

**Comment by GeneDer** [comment]
@rsong0606 are you using BentoML like the other issue? And can you share complete reproducible code so we can debug further? 

If the question is to set a larger rpc limit, you can probably try this env `RAY_max_grpc_message_size` https://github.com/ray-project/ray/blob/75161ab72f8d76c9365756b5d86f111582bbb42a/python/ray/_private/runtime_env/packaging.py#L36

And btw if that is the issue, then it should have be tagged core issue instead of serve issue.

**Comment by rsong0606** [comment]
@rsong0606 Thanks for your response. Well noted on the core issue instead of the serve issue.

I tried to set the RAY_max_grpc_message_size from 500* to 1000* still getting the error.

I am not using BentoML, I am building my application on a GCP instance, with Fastapi + Ray. My goal is to serve with Ray and handle concurrent requests from multiple user(s) simultaneously without crashing. And I am new to Ray. Thanks for your help again and I appreciate your valuable time. Please let me know if the code can be used for reproducible.

I have 3 Python files for running ray serve, I tried to bind all models into one application.

first is ray_deployment which takes a deployment decorator from Ray

```
RES_DIR = os.getenv('RES_DIR', '/default/path/to/imgs')
MODEL_DIR = os.getenv('MODEL_DIR', '')
ZH_TRANSLATOR = Translator(to_language="zh")
RESOURCE_DIR = os.getenv('RESOURCE_DIR', '/default/path/to/resources')

logger = logging.getLogger("ray.serve")

@serve.deployment(ray_actor_options={"num_gpus": 0.2})
class DictionaryLookUpDeployment:
    def __init__(self):
        self.handler = DictionaryLookUpHandler()

    async def __call__(self, request):
        final_output = self.handler.generate_and_concatenate(request.word, ZH_TRANSLATOR)
        return final_output

@serve.deployment(ray_actor_options={"num_gpus": 0.3})
class ImageHandlerDeployment:
    def __init__(self):
        self.handler = ImageHandler(folder_path=str(RES_DIR))

    async def __call__(self, target_language: str):
        image_path, description, translated_description = self.handler.process_image(ZH_TRANSLATOR)
        encoded_image = encode_image_to_base64(image_path)
        return {
            "image_path": encoded_image,
            "description": description,
            "translated_description": translated_description
        }

@serve.deployment(ray_actor_options={"num_gpus": 0.5})
class ChatbotDeployment:
    def __init__(self):
        self.loop = asyncio.get_running_loop()
        self.model_id = MODEL_DIR
        self.model = AutoModelForCausalLM.from_pretrained(self.model_id)
        self.tokenizer = AutoTokenizer.from_pretrained(self.model_id)

    async def __call__(self, user_input: str):
        streamer = TextIteratorStreamer(
            self.tokenizer, timeout=0, skip_prompt=True, skip_special_tokens=True
        )
        self.loop.run_in_executor(None, self.generate_text, user_input, streamer)
        return StreamingResponse(
            self.consume_streamer(streamer), media_type="text/plain"
        )

    def generate_text(self, prompt: str, streamer: TextIteratorStreamer):
        input_ids = self.tokenizer([prompt], return_tensors="pt").input_ids
        self.model.generate(input_ids, streamer=streamer, max_length=10000)

    async def consume_streamer(self, streamer: TextIteratorStreamer):
        while True:
            try:
                for token in streamer:
                    logger.info(f'Yielding token: "{token}"')
                    yield token
                break
            except Empty:
                await asyncio.sleep(0.001)

@serve.deployment(ray_actor_options={"num_gpus": 0.2})
class AudioEvaluatorDeployment:
    def __init__(self):
        self.handler = AudioEvaluator()

    async def __call__(self, user_audio_path: str, reference_audio_path: str):
        feedback = self.handler.evaluate_user_pronunciation(user_audio_path, reference_audio_path)
        return feedback

@serve.deployment
class DictionaryViewerDeployment:
    def __init__(self):
        self.handler = DictionaryViewer(str(RESOURCE_DIR))

    async def __call__(self):
        return self.handler.show_next_item()

@serve.deployment
class QuizHandlerDeployment:
    def __init__(self):
        self.handler = EnglishQuizHandler()

    async def __call__(self, request: str):
        return self.handler.check_answer(request)

@serve.deployment(ray_actor_options={"num_gpus": 0.2})
class AudioGeneratorDeployment:
    def __init__(self):
        self.handler = AudioGenerator(model_name="suno/bark-small")

    async def __call__(self, text: str, language: str):
        audio_array = self.handler.generate_audio(text, language)
        return audio_array

dictionary_lookup = DictionaryLookUpDeployment.bind()
image_handler = ImageHandlerDeployment.bind()
chatbot = ChatbotDeployment.bind()
audio_evaluator = AudioEvaluatorDeployment.bind()
dictionary_viewer = DictionaryViewerDeployment.bind()
quiz_handler = QuizHandlerDeployment.bind()
audio_generator = AudioGeneratorDeployment.bind()
```


Second is my main.py which runs the application

```
import os
import sys
from pathlib import Path

# Add the project directory to the Python path
PROJECT_DIR = Path(__file__).resolve().parent
sys.path.append(str(PROJECT_DIR))
os.environ["RAY_max_grpc_message_size"] = str(1000 * 1024 * 1024)

# Verify the environment variable is set correctly
print(f"RAY_max_grpc_message_size: {os.environ.get('RAY_max_grpc_message_size')}")

import ray
from ray import serve
from ray_deploy import dictionary_lookup, image_handler, chatbot, audio_evaluator, dictionary_viewer, quiz_handler, audio_generator
from fast_api import app, MyFastAPIDeployment  # Correctly import the FastAPI app and deployment class
# Initialize Ray and Ray Serve
ray.init(address="auto")
serve.start()




my_fastapi_deployment = MyFastAPIDeployment.bind(
    dictionary_lookup, image_handler, chatbot,
    audio_evaluator, dictionary_viewer, quiz_handler, audio_generator
)
serve.run(my_fastapi_deployment)

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
```


and the last one is my fast_api.py which builds different API calls and bind into Ray

the last part is taking fastapi into Ray

```
@serve.deployment
@serve.ingress(app)
class MyFastAPIDeployment:
    def __init__(self, dictionary_lookup_handle: DeploymentHandle, image_handler_handle: DeploymentHandle, chatbot_handle: DeploymentHandle,
                 audio_evaluator_handle: DeploymentHandle, dictionary_viewer_handle: DeploymentHandle, quiz_handler_handle: DeploymentHandle,
                 audio_generator_handle: DeploymentHandle):
        app.state.dictionary_lookup_handle = dictionary_lookup_handle
        app.state.image_handler_handle = image_handler_handle
        app.state.chatbot_handle = chatbot_handle
        app.state.audio_evaluator_handle = audio_evaluator_handle
        app.state.dictionary_viewer_handle = dictionary_viewer_handle
        app.state.quiz_handler_handle = quiz_handler_handle
        app.state.audio_generator_handle = audio_generator_handle
```


I was experimenting with this today, this code actually works, I am not sure if the autoscaling_config made it work here.


```
app = FastAPI()

class UserInput(BaseModel):
    text: str

@serve.deployment
@serve.ingress(app)
class APIIngress:
    def __init__(self, chatbot_handle: DeploymentHandle) -> None:
        self.handle = chatbot_handle

    @app.post("/chatbot")
    async def generate_response(self, user_input: UserInput):
        return await self.handle.generate_response.remote(user_input.text)

@serve.deployment(
    ray_actor_options={"num_gpus": 0.3},
    autoscaling_config={"min_replicas": 1, "max_replicas": 2},
)
class ChatbotDeployment:
    def __init__(self):
        self.chatbot = Chatbot("")

    def generate_response(self, user_input: str):
        return self.chatbot.generate_response(user_input)

entrypoint = APIIngress.bind(ChatbotDeployment.bind())

def get_app_chatbot():
    return entrypoint

if __name__ == "__main__":
    import ray
    ray.init(address="auto")
    serve.start()
    serve.run(get_app_chatbot())
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8001)
```

**Comment by rsong0606** [comment]
@GeneDer Hey Gene, 

I found another approach to handle multiple applications in one shot  and found the code for reproducible.

Currently I created bunch of python files like below



```
import re
import psutil
import time

import uvicorn
from vllm import LLM, SamplingParams
import torch
from fastapi import FastAPI, Request
from pydantic import BaseModel
from ray import serve
from ray.serve.handle import DeploymentHandle
import ray
class Chatbot:
    def __init__(self, model_dir):
        # Load the vLLM model with adjusted parameters
        self.llm = LLM(
            model=model_dir,
            quantization="gptq_marlin",
            dtype=torch.float16,  # Use mixed precision
            enforce_eager=True,  # Enforce eager mode to avoid CUDA graphs
            gpu_memory_utilization=0.5,
        )
        # Initialize conversation history
        self.conversation_history = []
        self.inference_count = 0
        # Create a sampling params object for vLLM
        self.sampling_params = SamplingParams(
            temperature=0.65,
            top_p=0.9,
            top_k=50,
            n=5,
            best_of=10,
            presence_penalty=0.9,
            frequency_penalty=0.7,
            repetition_penalty=1.3,
            length_penalty=1.0,
            max_tokens=128,
            min_tokens=50,
            ignore_eos=False,
            skip_special_tokens=True,
            include_stop_str_in_output=False,
            stop=["\n", "[INST]", "[/INST]"]
        )

    def generate_response(self, user_input):
        # Add user input to the conversation history
        cleaned_user_input = self.clean_text(user_input)
        self.conversation_history.append(cleaned_user_input)

        print("Updated conversation history after user input:", self.conversation_history)

        # Create a prompt with the latest user input encapsulated within instructional tokens
        prompt = f"<s>[INST] {cleaned_user_input} [/INST]"

        # Track memory usage and inference time
        process = psutil.Process()
        start_mem = process.memory_info().rss
        start_time = time.time()

        # Generate the model's response using vLLM
        outputs = self.llm.generate([prompt], self.sampling_params)

        # Stop the timer after the model inference
        end_time = time.time()

        # Calculate the inference time and memory usage
        inference_time = end_time - start_time
        end_mem = process.memory_info().rss
        memory_usage = end_mem - start_mem

        print(f"Inference Time: {inference_time:.2f} seconds")
        print(f"Memory Usage: {memory_usage / (1024 ** 2):.2f} MB")

        # Extract the generated text
        generated_text = outputs[0].outputs[0].text

        # Clean the generated text
        response = self.clean_generated_text(generated_text, prompt)

        # Add the cleaned response to the conversation history for context in future interactions
        self.conversation_history.append(response)

        print("Updated conversation history after model response:", self.conversation_history)
        self.inference_count += 1

        # Clear cached GPU memory after every 5 inferences
        if self.inference_count % 5 == 0:
            torch.cuda.empty_cache()
            print("Cleared GPU cache after 5 inferences.")
        return response

    @staticmethod
    def clean_text(text):
        # Remove non-ASCII characters
        text = re.sub(r'[^\x20-\x7E]', '', text)  # Only keep printable ASCII characters
        return text

    @staticmethod
    def clean_generated_text(generated_text, prompt):
        # Remove the prompt part from the generated text
        response = generated_text.replace(prompt, '').strip()

        # Capture everything before the first [/INST] marker
        response_before_first_end_token = response.split('[/INST]', 1)[0].strip()

        # Extract text between [INST] and [/INST] if present
        inst_content = re.search(r'\[INST\](.*?)\[\/INST\]', response_before_first_end_token)
        if inst_content:
            response_before_first_end_token = inst_content.group(1).strip()
        response_before_first_end_token = re.sub(r'[^\x20-\x7E]', '', response_before_first_end_token)
        return response_before_first_end_token
```

And bind with serve, FastAPI with unicorn to expose a api for my frontend application to interact with it.


```
app = FastAPI()

class UserInput(BaseModel):
    text: str

@serve.deployment
@serve.ingress(app)
class APIIngress:
    def __init__(self, chatbot_handle: DeploymentHandle) -> None:
        self.handle = chatbot_handle

    @app.post("/chatbot/")
    async def generate_response(self, user_input: UserInput):
        return await self.handle.generate_response.remote(user_input.text)

@serve.deployment(
    ray_actor_options={"num_gpus": 0.3},
    autoscaling_config={"min_replicas": 1, "max_replicas": 2},
)
class ChatbotDeployment:
    def __init__(self):
        self.chatbot = Chatbot("CNBOOMBOOM/quant-gptq-hivetalk-8bits-10k-v4")

    def generate_response(self, user_input: str):
        return self.chatbot.generate_response(user_input)

entrypoint = APIIngress.bind(ChatbotDeployment.bind())

def get_app_chatbot():
    return entrypoint

if __name__ == "__main__":
    ray.init(address="auto")
    serve.start()
    serve.run(get_app_chatbot())
```

Once running this code, it returned 

```
2024-05-31 01:58:57,619 INFO handle.py:126 -- Created DeploymentHandle '41s76z74' for Deployment(name='APIIngress', app='default').
2024-05-31 01:58:57,619 INFO api.py:584 -- Deployed app 'default' successfully.
```

serve status returned
```
proxies:
  76b7d7946a18e4f96c4f4873766d2f5d79b5d2a992b8c223abda0e26: HEALTHY
applications:
  default:
    status: RUNNING
    message: ''
    last_deployed_time_s: 1717117952.164286
    deployments:
      ChatbotDeployment:
        status: HEALTHY
        status_trigger: CONFIG_UPDATE_COMPLETED
        replica_states:
          RUNNING: 1
        message: ''
      APIIngress:
        status: HEALTHY
        status_trigger: CONFIG_UPDATE_COMPLETED
        replica_states:
          RUNNING: 1
        message: ''
target_capacity: null
```

But when I added
  `uvicorn.run(app, host="0.0.0.0", port=8000)`
it returns the error message same as in the link.

```
    raise RayServeException(
ray.serve.exceptions.RayServeException: `serve.get_replica_context()` may only be called from within a Ray Serve deployment.
```

I actually want to expose the api address to interact with my frontend application, does Ray has a way to expose that?  Hopefully the code I provide now can be used for reproducing

**Comment by rsong0606** [comment]
Next issue is to deploy multiple applications simultaneously.

This link is helpful, https://docs.ray.io/en/latest/serve/multi-app.html.  However, in my current config.yaml, I might have multiple route_prefix for each application

```
@serve.deployment
@serve.ingress(app)
class APIIngress:
    def __init__(self, user_management_handle: DeploymentHandle) -> None:
        self.handle = user_management_handle

    @app.post("/api/register")
    async def register(self, user: UserCreate):
        print(f"Received registration request: {user}")
        if await self.handle.get_user.remote(user.username):
            raise HTTPException(status_code=400, detail="Username already registered")
        await self.handle.create_user.remote(user.username, user.password, user.email)
        await self.handle.print_users.remote()  # Print users after registration
        return {"message": "User registered successfully"}

    @app.post("/api/login")
    async def login(self, user: User):
        print(f"Received login request: {user}")
        db_user = await self.handle.get_user.remote(user.username)
        if not db_user or not pwd_context.verify(user.password, db_user[2]):
            raise HTTPException(status_code=400, detail="Incorrect username or password")
        await self.handle.update_last_login.remote(user.username)
        await self.handle.print_users.remote()  # Print users after login
        return {"message": "Login successful"}

@serve.deployment
class UserManagementDeployment:
    def get_user(self, username: str):
        return get_user(username)

    def create_user(self, username: str, password: str, email: str):
        create_user(username, password, email)

    def update_last_login(self, username: str):
        update_last_login(username)

    def print_users(self):
        print_users()

entrypoint = APIIngress.bind(UserManagementDeployment.bind())

def get_app_user_management():
    return entrypoint

if __name__ == "__main__":
    import ray
    ray.init(address="auto")
    serve.start()
    serve.run(get_app_user_management())
```

Also I am using FastApi to serve as well. 

Attached are my config.yaml

```
proxy_location: EveryNode

http_options:
  host: 0.0.0.0
  port: 8001

grpc_options:
  port: 9000
  grpc_servicer_functions: []

logging_config:
  encoding: JSON
  log_level: INFO
  logs_dir: null
  enable_access_log: true

applications:
  - name: scenario_learning_app
    route_prefix: /api/next_phrase
    import_path: scenarios_phrase_learning/senarios_learning:get_app_scenario_learning
    runtime_env: {}
    deployments:
      - name: APIIngress
      - name: ScenarioLearningDeployment

  - name: translator_app
    route_prefix: /api/translate
    import_path: translator:get_app_translator
    runtime_env: {}
    deployments:
      - name: APIIngress
      - name: TranslatorDeployment

  - name: caption_translation_app
    route_prefix: /api/process_image/
    import_path: img_to_text/caption_translation:get_app_image_handler
    runtime_env: {}
    deployments:
      - name: APIIngress
      - name: ImageHandlerDeployment

  - name: generated_audio_app
    route_prefix: /api/generate_audio/en/
    import_path: audio/generated_audio:get_app_audio_generator
    runtime_env: {}
    deployments:
      - name: APIIngress
      - name: AudioGeneratorDeployment

  - name: user_audio_app
    route_prefix: /api/evaluatePronunciation
    import_path: audio/user_audio:get_app_audio_evaluator
    runtime_env: { }
    deployments:
      - name: APIIngress
      - name: AudioEvaluatorDeployment

  - name: sqllite_database_app
    route_prefix: /api/register
    import_path: database/sqllite_database:get_app_user_management
    runtime_env: { }
    deployments:
      - name: APIIngress
      - name: UserManagementDeployment

  - name: phrase_learning_app
    route_prefix: /api/next_phrase1
    import_path: phrase_learning/phrase_learning:get_app_dictionary_viewer
    runtime_env: { }
    deployments:
      - name: APIIngress
      - name: DictionaryViewerDeployment

  - name: english_quiz_app
    route_prefix: /api/start_quiz
    import_path: quiz/english_quiz:get_app_quiz_handler
    runtime_env: { }
    deployments:
      - name: APIIngress
      - name: QuizHandlerDeployment

  - name: scenario_selection_app
    route_prefix: /api/chat
    import_path: scenarios_chat/scenario_selection:get_app_chatbot
    runtime_env: { }
    deployments:
      - name: APIIngress
      - name: ChatbotDeployment

  - name: fetch_edict_app
    route_prefix: /api/edict
    import_path: speaking_from_dict/fetch_edict:get_app_ec_dict
    runtime_env: { }
    deployments:
      - name: APIIngress
      - name: ECDictDeployment

```

a few questions
1. How we could integrate with FastAPI, is my implementation in yaml is right?
2. Does this yaml file will deploy all applications simultaneously? which means in local host, we will have all deployments in parallel?
3. How we could handle one file contains multiple route_prefix?

My current way to handle this is to create master_script which to append each file by file.


`if __name__ == "__main__":
    processes = []

    p1 = multiprocessing.Process(target=run_scenario_learning)
    processes.append(p1)

    p2 = multiprocessing.Process(target=run_audio_evaluator)
    processes.append(p2)

    p3 = multiprocessing.Process(target=run_user_management)
    processes.append(p3)

    p4 = multiprocessing.Process(target=run_audio_generator)
    processes.append(p4)

    p5 = multiprocessing.Process(target=run_image_handler)
    processes.append(p5)

    p6 = multiprocessing.Process(target=run_phrase_learning_handler)
    processes.append(p6)

    p7 = multiprocessing.Process(target=run_chatbot_handler)
    processes.append(p7)

    p8 = multiprocessing.Process(target=run_dict_lookup_hander)
    processes.append(p8)

    p9 = multiprocessing.Process(target=run_translator_handler)
    processes.append(p9)

    for p in processes:
        p.start()

    for p in processes:
        p.join()`  

But not all functions are in parallel, only returning one of all.

**Comment by abrarsheikh** [comment]
closing this issue, from reading the thread it appears that the original issue got resolved.
